src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +68,89 @@ def install_given_reqs(
             ', '.join(to_install.keys()),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = []  # type: List[InstallationResult]
+
+    # store install arguments
+    install_args = _InstallArgs(
+        install_options=install_options,
+        global_options=global_options,
+        kwargs=dict(
+            root=root, home=home, prefix=prefix,
+            warn_script_location=warn_script_location,
+            use_user_site=use_user_site, pycompile=pycompile
+        )
+    )
 
     with indent_log():
-        for req_name, requirement in to_install.items():
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', req_name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            else:
-                uninstalled_pathset = None
-
-            try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
+        # first try to install in parallel
+        installed_pool = map_multiprocess_ordered(
+            partial(_single_install, install_args, in_subprocess=True),
+            requirements)
+
+        # check the results from the parallel installation,
+        # and fill-in missing installations or raise exception
+        for installed_req, requested_req in zip(installed_pool, requirements):
+            # if the requirement was not installed by the parallel pool,
+            # install serially here
+            if installed_req is None:
+                installed_req = _single_install(
+                    install_args, requested_req, in_subprocess=False)
+
+            if isinstance(installed_req, BaseException):
```

Comment:
Should this be `elif` so it does not continue after a serial install?

Reply:
Actually no, it checks if a package was skipped during the parallel install, if so it adds it, then second section will push into into the installed requirements list.